crates/matrix-sdk/src/room/mod.rs

```diff
@@ -90,6 +90,64 @@ pub use self::{
     messages::{Messages, MessagesOptions},
 };
 
+type DeduplicatedRequestMap<Key> = Mutex<BTreeMap<Key, Arc<Mutex<Result<(), ()>>>>>;
+
+/// Handler that properly deduplicates requests to the same endpoint, and will
+/// properly report error upwards in case the concurrent request failed.
+pub(crate) struct DeduplicatedRequestHandler<Key> {
+    inflight: DeduplicatedRequestMap<Key>,
+}
```

Comment:
Just realized that this is defined inside the `room` module, can you move it to a more appropriate place, given it's useful / used outside of here too? Maybe `matrix-sdk` should have a `util` module like `matrix-sdk-ui`.

Reply:
Sure. I'll make a `deduplicated_handler` module, I find utils mods very hard to discover, and not quite self-explanatory in general.